### PR TITLE
[MCOMPILER-592] - IndexOutOfBoundsException when project.build.output…

### PR DESCRIPTION
**Pull request by thijsrijpert**
…Timestamp is empty

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MCOMPILER) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MCOMPILER-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MCOMPILER-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by thijsrijpert** [comment]
From JIRA issue:

Due to a combination of plugins, it is possible for the project.build.outputTimestamp property to have an empty string as a value. This causes an IndexOutOfBoundsException on line 1242 when it tries to access the first character in the string.

This can be fixed by doing an isEmpty check before doing the charAt check. I will create an PR for this.

While this bug is caused by another plugin, I think this should be fixed in the compiler because the error returned (IndexOutOfBoundsException), is vague and hard to debug.

The workaround for this issue is setting the project.build.outputTimestamp property with an arbitrary value.

File: src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:1243]: I'd like to understand how the value can be empty: `<foo/>`? This basically can happen in any plugin using this value, no?

Comment (thijsrijpert) [reply src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:1243]: Yes, I think it would be set by another plugin. The project I was working on has a lot of plugins, so I would not know which plugin exactly sets this value. When I encountered this issue, the value was not set in the projects pom, but it was a non-null empty string when this line was run, indicating it was set somewhere in the compilation process.

Comment (michael-o) [reply src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:1243]: While I consider your fix as reasonable, I'd prefer to see the actual cause to be addressed as well...

Comment (thijsrijpert) [reply src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:1243]: I will look into it.

Comment (thijsrijpert) [reply src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:1243]: Hi I looked a bit further into this. 

In the (internal) project I was working on was using the jgitver maven extension. This extension has a field named COMMIT_ISO_TIMESTAMP which contains the timestamp of the commit. If the repository does not have any commits, this timestamp is empty. The project I was working on was mapping this timestamp to the project.build.outputTimestamp to this maven commit timestamp in the parent pom, which I missed before.

This caused the above mentioned compiler error.

Comment (michael-o) [reply src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:1243]: Have you raised an upstream issue?

Comment (michael-o) [reply src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:1243]: I'll give the maintainers some time to respond.

Comment (thijsrijpert) [reply src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:1243]: Hi,

On July 2th this MR has been approved by gnodet. Are we waiting for someone elses approval?

Regards,

Thijs Rijpert

**Review by olamy — APPROVED** [review]

**Pull request merged**
